packages/z3/z3.4.8.5/files/4468.patch

```diff
@@ -0,0 +1,267 @@
+shared library that could be loaded in runtime use
+`Dynlink.loadfile` (which used dlopen on posix machines underneath the
+hood).
+
+- libz3ml.a is the archived machine code for `z3native_stubs.c`, which
+is made by ocamlmklib: `ar rcs api/ml/libz3ml.a
+api/ml/z3native_stubs.o` it is needed to build statically linked
+binaries and libraries that use z3 bindings.
+
+- dllz3ml.so is the shared object that contains `z3native_stubs.o` as
+well as correct ldd entries for C++ and Z3 libraries to enable proper
+static and dynamic linking. The file is built with ocamlmklib on posix
+systems as
+```
+gcc -shared -o api/ml/dllz3ml.so api/ml/z3native_stubs.o -L. -lz3-static -lstdc++
+```
+
+It is used by `ocaml`, `ocamlrun`, and `ocamlc` to link z3 and c++
+code into the OCaml runtime and enables usage of z3 bindings in
+non-custom runtimes (default runtimes).
+
+The `dllz3ml.so` is usually installed in the stubs library in opam
+installation (`$(opam config var lib)/stublibs`), it is done
+automatically by `ocamlfind` so no special treatment is needed.
+
+Technical Details
+=================
+
+The patch itself is rather small. First of all, we have to use
+`-l<lib>` instead of `-cclib -l<lib>` in ocamlmklib since the latter
+will pass the options only to the ocaml{c,opt} linker and will not
+use the passed libraries when shared and non-shared versions of the
+bindings are built (libz3ml.a and dllz3ml.so). They were both missing
+either z3 code itself and ldd entries for stdc++ (and z3 if built not
+in --staticlib mode).
+
+Having stdc++ entry streamlines the compilation process and makes
+dynamic loading more resistant to the inclusion order.
+
+Finally, we had to add `-L.` to make sure that the built artifacts are
+correctly found by gcc.
+
+I specifically left the cygwin part of the code intact as I have no
+idea what the original author meant by this, neither do I use or
+tested this patch in the cygwin or mingw environemt. I think that this
+code is rather outdated and shouldn't really work. E.g., in the
+--staticlib mode adding z3linkdep (which is libz3-static.a) as an
+argument to `ocamlmklib` will yield the following broken archive
+```
+ar rcs api/ml/libz3ml.a libz3-static.a api/ml/z3native_stubs.o
+```
+and it is not allowed (or supported) to have .a in archives (though it
+doesn't really hurt as most of the systems will just ignore it).
+
+But otherwise, cygwin, mingw shall behave as they did (the only change
+that affects them is `-L.` which I believe should be benign).
+---
+ scripts/mk_util.py | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/scripts/mk_util.py b/scripts/mk_util.py
+index 1a95ee20c3..156be9ca2f 100644
+--- a/scripts/mk_util.py
++++ b/scripts/mk_util.py
+@@ -1957,7 +1957,7 @@ def mk_makefile(self, out):
+ 
+             OCAMLMKLIB = 'ocamlmklib'
+ 
+-            LIBZ3 = '-cclib -l' + z3link
++            LIBZ3 = '-l' + z3link + ' -lstdc++'
```

Comment:
```suggestion
+            LIBZ3 = '-l' + z3link + ' -lstdc++' + '-lgomp'
```

Reply:
@mseri, I can't really edit this PR (other than creating a PR to @floriangru master branch) so I just made the suggested change and any maintaner should be able to accep it). This should fix the linking error. 

Concerting the `conf-gomp` package, I am for skipping this step, as it is a part of the C++ toolchain and it is already in the dependencies of the z3 package.

Reply:
Mmh, it fails on macos due to missing gomp

```
#=== ERROR while compiling z3.4.8.5 ===========================================#
2100# context     2.0.7 | macos/x86_64 | ocaml-system.4.10.0 | file:///Users/travis/build/ocaml/opam-repository
2101# path        ~/.opam/ocaml-system/.opam-switch/build/z3.4.8.5
2102# command     ~/.opam/opam-init/hooks/sandbox.sh build make -C build -j 1
2103# exit-code   2
2104# env-file    ~/.opam/log/z3-93566-60459a.env
2105# output-file ~/.opam/log/z3-93566-60459a.out
2106### output ###
2107# 206 |   let compare = Pervasives.compare
2108# [...]
2109# Alert deprecated: module Stdlib.Pervasives
2110# Use Stdlib instead.
2111# 
2112# If you need to stay compatible with OCaml < 4.07, you can use the 
2113# stdlib-shims library: https://github.com/ocaml/stdlib-shims
2114# cp libz3.a libz3-static.a
2115# ocamlfind ocamlc -package num  -ccopt "-D_MP_INTERNAL -DNDEBUG -D_EXTERNAL_RELEASE   -fvisibility=hidden -c -mfpmath=sse -msse -msse2 -D_NO_OMP_ -O3 -Wno-unknown-pragmas -Wno-overloaded-virtual -Wno-unused-value -fPIC -I /usr/local/lib/ocaml -I ../src/api -I ../src/api/ml -o api/ml/z3native_stubs.o" -c ../src/api/ml/z3native_stubs.c
2116# ocamlmklib -o api/ml/z3ml -I api/ml -L. api/ml/z3native_stubs.o api/ml/z3enums.cmo api/ml/z3native.cmo api/ml/z3.cmo  -lz3-static -lstdc++ -lgomp 
2117# ld: library not found for -lgomp
2118# clang: error: linker command failed with exit code 1 (use -v to see invocation)
2119# make: *** [api/ml/z3ml.cma] Error 2
```